CMakeLists.txt

```diff
@@ -974,4 +984,53 @@ else()
             endif()
         endif()
     endif()
+
+    if (BUILD_SHARED_LIB STREQUAL "ON")
+        if(CRYPTO STREQUAL "mbedtls")
+            SET(CRYPTO_DEPS "-lmbedtls -lmbedx509 -lmbedcrypto")
+            add_library(${LIB_NAME} SHARED 
+                                            $<TARGET_OBJECTS:cryptlib_${CRYPTO}>
```

Comment:
This is open for debate, but as a shared library `libspdm.so` should only include the files in `libspdm/library/`
- spdm_common_lib
- spdm_crypt_lib
- spdm_requester_lib
- spdm_responder_lib
- spdm_secured_message_lib
- spdm_transport_mctp_lib
- spdm_transport_pcidoe_lib

Functionality beyond that would go in a separate shared library or be statically linked into the executable binary.

Reply:
I agree with your comments, but in this scenario we are trying to build a functional shared library that can be included and used directly in any libspdm integration. So I think debugging code, memory allocation, cryptographic library hooks, rng generator and time management should be part of the built shared library in my opinion. 
But you guys are the owners, so I can also create two shared libraries if you think that works better, like libspdm_main.so with what you mentioned and libspdm_platform.so that includes all the required primitives to be able to instantiate libspdm in an integration.
What is your opinion?

Reply:
Why not use 1:1 mapping for static lib and dynamic lib? Why we have to create 1 .so file including everything?

Reply:
Hi, we want to package libspdm as a regular shared library on linux distributions, for future work around SPDM attestation happening in a regular linux distribution (libspdm can act as either responder or requester, depending on target platform and usecase).
To that end we want to update the libspdm project to have an option to build a shared library like many other libraries in a linux operating system.
I understand your concern regarding need for customization of different components of a libspdm target, such as custom crypto library or HAL.
To that end I will create a new commit with a 1:1 mapping for static lib to shared lib, with an overarching libspdm.so shared library that depends on all the other libspdm libraries to run (using TARGET_LINK_LIBRARIES in CMakeLists.txt). 
This way, from the user point of view, a simple linking with libspdm.so is enough to compile the program, as long as all the other libraries are built and available in the system to use at runtime, as libspdm.so itself will internally depend on the other libraries.
Does that sound ok?

Reply:
Thanks @uvinrg . Sounds good to me to have 1:1 mapping.

BTW: For Linux use case, do you anticipate requester only? or responder as well?
I have confidence for the requester, such as attestation use case. That is standard CMA flow.
But I am not sure about the responder, which has more strict requirement on security (e.g. private key management) and platform (e.g. measurement reporting, watchdog for heartbeat). It would be good to know what your thought is.

Reply:
Hi, while we initially target requester usecase, responder as a shared library will be added later as well, as it is useful even if just for testing a system. Regarding security, there is only so much a library, shared or statically linked can do to protect its private key. Security of the running environment of the responder should be provided by the hardware platform and the operating system, so we don't see the shared library linking as an issue for the responder.
For the 1:1 mapping I have an issue as CMake does not allow both a shared and static taget with the same name to be generated. As such, if shared libraries are enabled, I have to disable building of the static libraries, which will also disable the building of the unit_tests, as they rely on the static libraries. Or change the name of the shared libraries, for example libspdm_requester_lib.a and libspdm_requester_lib_sh.so
What is your opinion?

Reply:
> Or change the name of the shared libraries, for example libspdm_requester_lib.a and libspdm_requester_lib_sh.so.

Yes, appending the `_sh` suffix should be fine.

Reply:
Yeah I'm fine with (at least) two libraries. The platform library would include
- debuglib
- malloclib
- memlib
- rnglib
- platformlib

The cryptography portion could either be part of the platform library or its own library.

Note that some of these are considered sample code, and since the intent is that they be used as production code they should be cleaned up and hardened in subsequent pull requests.